substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVLineRecord.java

```diff
@@ -0,0 +1,194 @@
+        }
+        return pos;
+    }
+
+    private int computeHeader(byte[] buffer, int initialPos) {
+
+        if (buffer == null) {
+            return initialPos + LINE_RECORD_HEADER_SIZE;
+        }
+
+        assert symbolName != null;
+        int pos = initialPos;
+
+        /* Emit addr:section relocation records. */
+        cvDebugInfo.getCVSymbolSection().markRelocationSite(pos, ObjectFile.RelocationKind.SECREL_4, symbolName, false, 1L);
+        pos = CVUtil.putInt(0, buffer, pos);
+        cvDebugInfo.getCVSymbolSection().markRelocationSite(pos, ObjectFile.RelocationKind.SECTION_2, symbolName, false, 1L);
+        pos = CVUtil.putShort((short) 0, buffer, pos);
+
+        /* Emit flags. */
+        pos = CVUtil.putShort(CB_HAS_NO_COLUMNS_FLAG, buffer, pos);
+
+        /* Length of this chunk in object file (= highAddr since it's zero based. */
+        int length = fileBlocks.isEmpty() ? 0 : fileBlocks.get(fileBlocks.size() - 1).getHighAddr();
+        pos = CVUtil.putInt(length, buffer, pos);
+        return pos;
+    }
+
+    boolean isEmpty() {
+        return fileBlocks.isEmpty();
+    }
+
+    @Override
+    public String toString() {
+        return String.format("CVLineRecord(type=0x%04x pos=0x%05x size=0x%d)", type, recordStartPosition, fileBlocks.size());
+    }
+
+    /*
+     * FileBlock is a section of contiguous code in a compilation unit, associated with a single
+     * source file. If a function includes inlined code, that code needs its own FileBlock,
+     * surrounded by FileBlocks describing the enclosing source file. A FileBlock consists of a list
+     * of LineEntries.
+     */
+    private static class FileBlock {
+
+        /* Fileblock header: fileId (4 bytes) lineEntry count (4 bytes) tablesize (4 bytes) */
+        static final int FILE_BLOCK_HEADER_SIZE = Integer.BYTES * 3;
+
+        private ArrayList<LineEntry> lineEntries = new ArrayList<>(DEFAULT_LINE_ENTRY_COUNT);
+        private int fileId;
+
+        FileBlock(int fileId) {
+            this.fileId = fileId;
+        }
+
+        void addEntry(LineEntry le) {
+            lineEntries.add(le);
+        }
+
+        int computeContents(byte[] buffer, int initialPos) {
+            if (buffer == null) {
+                return computeSize(initialPos);
+            }
+            int pos = initialPos;
+            pos = CVUtil.putInt(fileId, buffer, pos);
+            pos = CVUtil.putInt(lineEntries.size(), buffer, pos);
+            pos = CVUtil.putInt(computeSize(0), buffer, pos);
+            for (LineEntry lineEntry : lineEntries) {
+                pos = lineEntry.computeContents(buffer, pos);
+            }
+            return pos;
+        }
+
+        int computeSize(int initialPos) {
+            return initialPos + FILE_BLOCK_HEADER_SIZE + LineEntry.LINE_ENTRY_SIZE * lineEntries.size();
+        }
+
+        int getHighAddr() {
+            return lineEntries.isEmpty() ? 0 : lineEntries.get(lineEntries.size() - 1).addr;
```

Comment:
Similarly here. Since the builder already ensures that `lineEntries` is not empty, it's not a problem at the moment, but I think `0` would be the wrong value if only the last file block was empty.

So, again, it might be better to simply assert here that `lineEntries` is not empty and expect the caller to take care of it.
```suggestion
            assert !lineEntries.isEmpty();
            return lineEntries.get(lineEntries.size() - 1).addr;
```

Reply:
Done.